lib/Arch/X86_C.cpp

```diff
@@ -0,0 +1,152 @@
+// registers. This includes parameters/arguments, return values, and the return
+// stack pointer.
+void X86_C::AllocateSignature(FunctionDecl &fdecl, const llvm::Function &func) {
+  // Bind return values first to see if we have injected an sret into the
+  // parameter list. Then, bind the parameters. It is important that we bind the
+  // return values before the parameters in case we inject an sret.
+  bool injected_sret = false;
+  fdecl.returns = BindReturnValues(func, injected_sret);
+  fdecl.params = BindParameters(func, injected_sret);
+  BindReturnStackPointer(fdecl, func);
+}
+
+// The return stack pointer describes where the stack will be upon return from
+// the function in terms of the registers of the current function. For x86_C,
+// this is usually ESP + 4 since that is where the return address is stored.
+void X86_C::BindReturnStackPointer(
+      FunctionDecl &fdecl, const llvm::Function &func) {
+  // Check if the first argument is an sret. If it is, then by the X86_C ABI,
+  // the callee is responbile for returning said sret argument in %eax and
+  // cleaning up the sret argument with a `ret 4`. This changes the
+  // return_stack_pointer offset because it will now be 4 bytes higher than we
+  // thought.
+  //
+  // However, even if there is sret on the second argument as well, we do not
+  // need to worry about this. For some reason the callee is only responsible
+  // for cleaning up the case where an sret argument is passed in as the first
+  // argument.
+  if (func.hasParamAttribute(0, llvm::Attribute::StructRet)) {
+    fdecl.return_stack_pointer_offset = 8;
+  } else {
+    fdecl.return_stack_pointer_offset = 4;
+  }
+
+  fdecl.return_stack_pointer = arch->RegisterByName("ESP");
+}
+
+std::vector<anvill::ValueDecl> X86_C::BindReturnValues(
+    const llvm::Function &function, bool &injected_sret) {
+  std::vector<anvill::ValueDecl> return_value_declarations;
+  anvill::ValueDecl value_declaration = {};
+  injected_sret = false;
+
+  // If there is an sret parameter then it is a special case. For the X86_C ABI,
+  // the sret parameters are guarenteed the be in %eax. In this case, we can
+  // assume the actual return value of the function will be the sret struct
+  // pointer.
+  for (auto arg = function.arg_begin(); arg != function.arg_end(); arg++) {
+    if (arg->hasStructRetAttr()) {
+      value_declaration.type = arg->getType();
+      value_declaration.reg = arch->RegisterByName("EAX");
+      return_value_declarations.push_back(value_declaration);
+      return return_value_declarations;
+    }
+  }
+
+  llvm::Type *ret_type = function.getReturnType();
+  value_declaration.type = ret_type;
+
+  switch (ret_type->getTypeID()) {
+    case llvm::Type::IntegerTyID:
+    case llvm::Type::PointerTyID: {
+      // Allocate EAX for an integer or pointer
+      value_declaration.reg = arch->RegisterByName("EAX");
+      break;
+    }
+    case llvm::Type::FloatTyID: {
```

Comment:
`double`, `log double` can be handled here too.

Reply:
mmx types can also be handled by a simple case of looking for the `MM0` register.

Reply:
fixed 028b2a5